sys/riotboot/slot.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *                    Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_riotboot_slot
+ * @{
+ *
+ * @file
+ * @brief       Slot management functions
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Francisco Acosta <francisco.acosta@inria.fr>
+ *
+ * @}
+ */
+#include <string.h>
+
+#include "cpu.h"
+#include "riotboot/slot.h"
+#include "riotboot/hdr.h"
+
+/*
+ * Store the start addresses of each slot.
+ * Take into account that CPU_FLASH_BASE represents the starting
+ * address of the bootloader, thus the header is located after the
+ * space reserved to the bootloader.
+ */
+const riotboot_hdr_t * const riotboot_slots[] = {
+    (riotboot_hdr_t*)(CPU_FLASH_BASE + SLOT0_OFFSET),   /* First slot address -> firmware image */
+};
+
+/* Calculate the number of slots */
+const unsigned riotboot_slot_numof = sizeof(riotboot_slots) / sizeof(riotboot_hdr_t*);
+
+static void _riotboot_slot_jump_to_image(const riotboot_hdr_t *hdr)
+{
+    cpu_jump_to_image(hdr->start_addr);
+}
+
+int riotboot_slot_current(void)
+{
+    uint32_t base_addr = cpu_get_image_baseaddr();
+
+    for (unsigned i = 0; i < riotboot_slot_numof; i++) {
+        const riotboot_hdr_t *hdr = riotboot_slot_get_hdr(i);
+        if (base_addr == hdr->start_addr) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+void riotboot_slot_jump(unsigned slot)
+{
+    _riotboot_slot_jump_to_image(riotboot_slot_get_hdr(slot));
+}
+
+uint32_t riotboot_slot_get_image_startaddr(unsigned slot)
+{
+    return riotboot_slot_get_hdr(slot)->start_addr;
```

Comment:
I'm actually unsure that `hdr` can be a NULL pointer. And not because it will always be there, but because we're returning an adress from flash, which I'm not sure it can be `NULL`.

IIRC I think this line is a debug leftover so it can be removed. Which other function makes use of it?

Reply:
just `riotboot_slot_dump_addrs`, sorry for the slightly confusing phrasing.

Reply:
I just checked the removal of the `hdr == NULL` and it still works properly.

Reply:
Merged, thanks